### PR TITLE
added generic  type for multiple type support

### PR DESCRIPTION
**Pull request by ammokhov**
*Issue #, if available:*

*Description of changes:*

added a generic type for multi type support. 

Example:

```json
"properties": {
    "property1": {
        "type": ["string", "object"]
    }
}
```
generated `ResourceModel.java`:
```java
@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class ResourceModel {
    @JsonProperty("property1")
    private Object property1;
    ...
```

test cases:
* 1  [CREATE] with escaped string `property1` -> [UPDATE] with modified escaped string `property1`
* 2 [CREATE] with escaped string `property1` -> [UPDATE] with json object `property1` 
* 3 [CREATE] with json object `property1` -> [UPDATE] with modified json `property1`
* 4 [CREATE] with list input for `property1` -> Failed with error `Model validation failed (#/property1: #: no subschema matched out of the total 2 subschemas) #/property1: expected type: JSONObject, found: JSONArray (#/property1) #/property1: expected type: String, found: JSONArray (#/property1)` 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rjlohan** [comment]
I think we actually want to support a fix for this issue: https://github.com/aws-cloudformation/cloudformation-cli/issues/242

`oneOf` is the correct schema structure for a union type.

File: src/rpdk/core/jsonutils/resolver.py [hunk 1]
```diff
@@ -104,7 +113,7 @@ def _schema_to_lang_type(self, property_schema):
             return self._get_array_lang_type(property_schema)
         if schema_type == "object":
             return self._get_object_lang_type(property_schema)
-        return self._get_primitive_lang_type(schema_type)
+        return self._get_primitive_lang_type(convert_list_to_object(schema_type))
```
Comment (wbingli) [inline src/rpdk/core/jsonutils/resolver.py:116]: It's more like hacking into the primitive type, but actually we should have a different type. e.g. 

```
if isinstance(schema_type, list)
       return self._get_multiple_lang_type(property_schema)

def _get_multiple_lang_type(self, property_schema):
 return ResolvedType(ContainerType.MULTIPLE, property_schema)

```

Comment (ammokhov) [reply src/rpdk/core/jsonutils/resolver.py]: done

File: src/rpdk/core/project.py [hunk 2]
```diff
@@ -434,7 +438,10 @@ def _set_docs_properties(
 
         prop_type = prop.get("type", "object")
 
-        if prop_type in BASIC_TYPE_MAPPINGS:
+        if isinstance(prop_type, list):
+            mapped = MULTIPLE_TYPE_MAPPINGS[MULTIPLE]
+            prop["jsontype"] = prop["yamltype"] = prop["longformtype"] = mapped
```
Comment (wbingli) [inline src/rpdk/core/project.py:443]: Here is to generate the doc.  We actually could show all the types in the doc, it's different as to generate the Java model.  If we only generate 'Object' for now,  not need the extra `MULTIPLE_TYPE_MAPPINGS[MULTIPLE]`

File: src/rpdk/core/jsonutils/resolver.py [hunk 3]
```diff
@@ -116,6 +122,10 @@ def _get_array_container_type(property_schema):
             return ContainerType.LIST
         return ContainerType.SET
 
+    @staticmethod
+    def _get_multiple_lang_type():
+        return ResolvedType(ContainerType.MULTIPLE, MULTIPLE)
```
Comment (wbingli) [inline src/rpdk/core/jsonutils/resolver.py:127]: It's better to pass the multiple types information, `return ResolvedType(ContainerType.MULTIPLE, schema_type)`

File: src/rpdk/core/jsonutils/resolver.py [hunk 4]
```diff
@@ -109,7 +109,7 @@ def _schema_to_lang_type(self, property_schema):
         if isinstance(
             schema_type, list
         ):  # generate a generic type Object which will be casted on the client side
-            return self._get_multiple_lang_type()
+            return self._get_multiple_lang_type(MULTIPLE)
```
Comment (wbingli) [inline src/rpdk/core/jsonutils/resolver.py:112]: Should it be `self._get_multiple_lang_type(schema_type)`?

**Comment by aygold92** [comment]
what is the use case for this? As far as I can tell, an indvidual property should never have two different possible types. This would be horrible API design and also does not make sense to map to a language model (which is why you need to cast it here)

**Comment by benkehoe** [comment]
There are plenty of places that you might want multiple types. A common one is to allow a nominally list-of-strings type to be given as string directly when there's only one element. The AWS IAM policy language supports this, and is embedded in a number of CloudFormation resources.

However, I agree with @rjlohan that instead of extending the JSON Schema spec, `oneOf` is the right way to support this in the schema.

**Comment by aygold92** [comment]
> There are plenty of places that you might want multiple types. A common one is to allow a nominally list-of-strings type to be given as string directly when there's only one element. The AWS IAM policy language supports this, and is embedded in a number of CloudFormation resources.

true, if we want to fully model an IAM policy, we would have to either allow this or allow multiple types declared using oneOf.  It probably would make the most to allow either/or.  This PR needs to edit this validation and merge the types into an array here https://github.com/aws-cloudformation/cloudformation-cli/blob/master/src/rpdk/core/jsonutils/utils.py#L154
It wouldn't make sense to allow multiple types to be declared in one way but not using oneOf

**Comment by aygold92** [comment]
> This PR needs to edit this validation and merge the types into an array here https://github.com/aws-cloudformation/cloudformation-cli/blob/master/src/rpdk/core/jsonutils/utils.py#L154
> It wouldn't make sense to allow multiple types to be declared in one way but not using oneOf

track that in https://github.com/aws-cloudformation/cloudformation-cli/issues/242

**Comment by srujithpoondla03** [comment]
> > There are plenty of places that you might want multiple types. A common one is to allow a nominally list-of-strings type to be given as string directly when there's only one element. The AWS IAM policy language supports this, and is embedded in a number of CloudFormation resources.
> 
> true, if we want to fully model an IAM policy, we would have to either allow this or allow multiple types declared using oneOf. It probably would make the most to allow either/or. This PR needs to edit this validation and merge the types into an array here https://github.com/aws-cloudformation/cloudformation-cli/blob/master/src/rpdk/core/jsonutils/utils.py#L154
> It wouldn't make sense to allow multiple types to be declared in one way but not using oneOf

As Json schema already supports the functionality of using oneOf to specify multiple types, I think we should utilize the same and that makes our life easier in terms of validating the schema and template in later stages. If we support this new generic type it opens multiple options in authoring the schema where schema authors could confuse with the original oneOf usage.

**Pull request merged**